server/routes/index.js

```diff
@@ -383,10 +383,32 @@ router.get('/dashboard/:study/:subject', ensurePermission, function (req, res) {
                           var encrypted = createHash('sha256').update(collectionName).digest('hex');
                           var varName = default_config[configItem].variable;
                           var escapedVarName = encodeURIComponent(varName).replace(/\./g, '%2E');
-                          var query = '[{ $project: {_id: 0, day: 1, "' + escapedVarName + '" : "$' + varName + '"}}]';
-                          var data = yield mongoData.collection(encrypted.toString()).aggregate(eval(query)).toArray();
-                          var queryForStat = '[{ $match : {"' + escapedVarName + '" : { $ne: "" }}},{ $group : { _id: "$null", min: {$min : "$' + escapedVarName + '"}, max: {$max : "$' + escapedVarName + '"}, mean: {$avg : "$' + escapedVarName + '"}}}]';
-                          var stat = yield mongoData.collection(encrypted.toString()).aggregate(eval(queryForStat)).toArray();
+                          const query = [{
+                            $project: { _id: 0, day: 1, [escapedVarName]: `$${varName}` }
+                          }];
+                          var data = yield mongoData.collection(encrypted.toString()).aggregate(query).toArray();
+                          const queryForStat = [
```

Comment:
I believe only `queryForStat` has been changed. I know our predecessors didn't break the lines and so it is being difficult to compare easily. Can you comment on the line that changed?

Reply:
Does [this](https://github.com/PREDICT-DPACC/dpdash/blob/34b81685a433aac8618d44e92b41b45686431e73/server/routes/index.js#L386-L411) block of code exist in four places in `index.js` :o ? If they are identical, should we open an issue so that we can write a function in the future to replace them all? Otherwise, as you can see, you have had to update in four places!

Reply:
`query` and `queryForStat` both changed. Previously these were being written as strings and passed to MongoDB's `eval` function, which was deprecated in 3.x and removed entirely in 4.4. That function takes in JavaScript and evaluates it in MongoDB. 

However, there was no need to use `eval` because the MongoDB node.js driver lets you pass JS objects directly as query objects. My guess is that the previous developer didn't know they could do the following transformations:
```js
"' + escapedVarName + '" : "$' + varName + '"
```
to
```js
[escapedVarName]: `$${varName}`
```
Surrounding `escapedVarName` in `[]` means that the variable will be evaluated and its resulting string will be used as the key, instead of evaluating as a key called "`escapedVarName`" which is what would happen without the `[]`. And `$${varName}` in backticks is shorthand for `'"$' + varName + '"'`.

You will then see that I pass `query` and `queryForStat` directly to `.aggregate()` instead of wrapping them inside of `eval()` first.

Reply:
> Does [this](https://github.com/PREDICT-DPACC/dpdash/blob/34b81685a433aac8618d44e92b41b45686431e73/server/routes/index.js#L386-L411) block of code exist in four places in `index.js` :o ? If they are identical, should we open an issue so that we can write a function in the future to replace them all? Otherwise, as you can see, you have had to update in four places!

Yes, I noticed that too. There are many places in this code that could be more modularized like that. I didn't bother yet since it's more of a "code cleaning" task than something dealing with mission critical functionality. but probably worth opening an issue.